sklearn/metrics/_regression.py

```diff
@@ -468,14 +503,22 @@ def explained_variance_score(y_true, y_pred, *,
     denominator = np.average((y_true - y_true_avg) ** 2,
                              weights=sample_weight, axis=0)
 
-    nonzero_numerator = numerator != 0
-    nonzero_denominator = denominator != 0
-    valid_score = nonzero_numerator & nonzero_denominator
-    output_scores = np.ones(y_true.shape[1])
+    if not fix_when_y_true_is_constant:
+        # Standard formula, that may lead to NaN or -Inf
+        output_scores = 1 - (numerator / denominator)
+    else:
+        nonzero_denominator = denominator != 0
+        nonzero_numerator = numerator != 0
+        # Default = Zero Numerator = perfect predictions. Set to 1.0
+        output_scores = np.ones([y_true.shape[1]])
+        # Non-zero Numerator and Non-zero Denominator: use the formula
+        valid_score = nonzero_denominator & nonzero_numerator
+        output_scores[valid_score] = 1 - (numerator[valid_score] /
+                                          denominator[valid_score])
+        # Non-zero Numerator and Zero Denominator:
+        # arbitrary set to 0.0 to avoid -inf scores
+        output_scores[nonzero_numerator & ~nonzero_denominator] = 0.
```

Comment:
Could you factorize the duplicated logic (present in this block and others) in a dedicated private function?